utils/src/main/java/software/amazon/awssdk/utils/cache/NonBlocking.java

```diff
@@ -37,80 +46,204 @@
 public class NonBlocking implements CachedSupplier.PrefetchStrategy {
     private static final Logger log = Logger.loggerFor(NonBlocking.class);
 
+    /**
+     * The maximum number of concurrent refreshes before we start logging warnings and skipping refreshes.
+     */
+    private static final int MAX_CONCURRENT_REFRESHES = 100;
+
+    /**
+     * The {@link Random} instance used for calculating jitter of the background prefetches.
+     */
+    private static final Random JITTER_RANDOM = new Random();
+
+    /**
+     * Thread used to kick off refreshes during the prefetch window. This does not do the actual refreshing. That's left for
+     * the {@link #EXECUTOR}.
+     */
+    private static final ScheduledThreadPoolExecutor SCHEDULER =
+        new ScheduledThreadPoolExecutor(1, new ThreadFactoryBuilder().daemonThreads(true).build());
+
+    /**
+     * Threads used to do the actual work of refreshing the values (because the cached supplier might block, so we don't
+     * want the work to be done by a small thread pool). This executor is created as unbounded, but we start complaining and
+     * skipping refreshes when there are more than {@link #MAX_CONCURRENT_REFRESHES} running.
+     */
+    private static final ThreadPoolExecutor EXECUTOR =
+        new ThreadPoolExecutor(0, Integer.MAX_VALUE,
+                               60L, TimeUnit.SECONDS,
+                               new SynchronousQueue<>(),
+                               new ThreadFactoryBuilder().daemonThreads(true).build());
+
+    /**
+     * An incrementing number, used to uniquely identify an instance of NonBlocking in the {@link #asyncThreadName}.
+     */
+    private static final AtomicLong THREAD_NUMBER = new AtomicLong(0);
```

Comment:
Perhaps better named "INSTANCE_NUMBER" as it tracks which instance of NonBlocking this is, separate from which thread it may be running on (which may vary over time)

Reply:
Fixed.